cpp/src/arrow/flight/flight_sql/CMakeLists.txt

```diff
@@ -0,0 +1,109 @@
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+add_custom_target(arrow_flight_sql)
+
+arrow_install_all_headers("arrow/flight/flight_sql")
+
+set(FLIGHT_SQL_PROTO_PATH "${ARROW_SOURCE_DIR}/../format")
+set(FLIGHT_SQL_PROTO ${ARROW_SOURCE_DIR}/../format/FlightSql.proto)
+
+set(FLIGHT_SQL_GENERATED_PROTO_FILES "${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.cc"
+                                     "${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.h")
+
+set(PROTO_DEPENDS ${FLIGHT_SQL_PROTO} ${ARROW_PROTOBUF_LIBPROTOBUF})
+
+add_custom_command(OUTPUT ${FLIGHT_SQL_GENERATED_PROTO_FILES}
+                   COMMAND ${ARROW_PROTOBUF_PROTOC} "-I${FLIGHT_SQL_PROTO_PATH}"
+                           "--cpp_out=${CMAKE_CURRENT_BINARY_DIR}" "${FLIGHT_SQL_PROTO}"
+                   DEPENDS ${PROTO_DEPENDS})
+
+set_source_files_properties(${FLIGHT_SQL_GENERATED_PROTO_FILES} PROPERTIES GENERATED TRUE)
+
+add_custom_target(flight_sql_protobuf_gen ALL DEPENDS ${FLIGHT_SQL_GENERATED_PROTO_FILES})
+
+set(ARROW_FLIGHT_SQL_SRCS server.cc client.cc)
+
+add_arrow_lib(arrow_flight_sql
+              CMAKE_PACKAGE_NAME
+              ArrowFlightSql
+              PKG_CONFIG_NAME
+              arrow-flight-sql
+              OUTPUTS
+              ARROW_FLIGHT_SQL_LIBRARIES
+              SOURCES
+              ${ARROW_FLIGHT_SQL_SRCS}
+              PRECOMPILED_HEADERS
+              "$<$<COMPILE_LANGUAGE:CXX>:arrow/flight/flight_sql/pch.h>"
+              DEPENDENCIES
+              flight_sql_protobuf_gen
+              SHARED_LINK_FLAGS
+              ${ARROW_VERSION_SCRIPT_FLAGS} # Defined in cpp/arrow/CMakeLists.txt
+              SHARED_LINK_LIBS
+              arrow_flight_shared
+              STATIC_LINK_LIBS
+              arrow_flight_static)
+
+if(ARROW_TEST_LINKAGE STREQUAL "static")
+  set(ARROW_FLIGHT_TEST_LINK_LIBS
+      arrow_flight_static arrow_flight_testing_static ${ARROW_FLIGHT_STATIC_LINK_LIBS}
+      ${ARROW_TEST_LINK_LIBS})
+  set(ARROW_FLIGHT_SQL_TEST_LINK_LIBS
+      arrow_flight_sql_static arrow_flight_testing_static
+      ${ARROW_FLIGHT_STATIC_LINK_LIBS} ${ARROW_TEST_LINK_LIBS})
+else()
+  set(ARROW_FLIGHT_TEST_LINK_LIBS arrow_flight_shared arrow_flight_testing_shared
+                                  ${ARROW_TEST_LINK_LIBS})
+  set(ARROW_FLIGHT_SQL_TEST_LINK_LIBS arrow_flight_sql_shared arrow_flight_testing_shared
+                                      ${ARROW_TEST_LINK_LIBS})
+endif()
+
+add_executable(flight_sql_test_app test_app_cli.cc
+                                   ${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.cc)
```

Comment:
I tried this patch, but it don't solve the ODR violation and tests fail...
I also noticed that if I just change `builtin-baseline` key on vcpkg.json the tests already fail in the current workaround. Still investigating this

Reply:
Hmm, ok, I'll take another look too when I can.

We might have to go with one of these options:
- Only allow dynamically linking libarrow_flight_sql when Protobuf is also dynamically linked (hence under vcpkg we would have to statically link everything)
- Remove libarrow_flight_sql and keep everything in libarrow_flight.so (however if one of the tests also needs to link Protobuf this won't fix things)

Reply:
Actually, the first suggestion is probably the only way we can solve things. Presumably what's currently happening is that:

- We have only libprotobuf.a (due to vcpkg),
- We link it into both libarrow_flight.so and libarrow_flight_sql.so,
- And we link both into the tests.

Hence we're violating ODR by definition. The only solution is that if we have static protobuf, then we can only build static Flight libraries (since then we won't link libprotobuf.a into libarrow_flight.a, expecting the user to link it themselves).

Reply:
Well, now with the latest PR this no longer fails for me…

Reply:
Though the tests do all fail in general, so maybe it's just a different symptom.

Reply:
Yeah - linking tests statically fixes it. Ok, I'm going to continue to dig into this, but I'm thinking the only solution will be to disallow dynamic linking of libarrow_flight if Protobuf and/or gRPC are statically linked.

Reply:
@rafael-telles how about this? This links the tests statically if we have static protobuf. The .so is still a landmine for users, though...

```diff
--- a/cpp/src/arrow/flight/CMakeLists.txt
+++ b/cpp/src/arrow/flight/CMakeLists.txt
@@ -25,7 +25,18 @@ if(WIN32)
   list(APPEND ARROW_FLIGHT_LINK_LIBS ws2_32.lib)
 endif()
 
-if(ARROW_TEST_LINKAGE STREQUAL "static")
+set(ARROW_FLIGHT_TEST_LINKAGE "${ARROW_TEST_LINKAGE}" PARENT_SCOPE)
+if (Protobuf_USE_STATIC_LIBS)
+  message(STATUS "Linking Arrow Flight tests statically due to static Protobuf")
+  set(ARROW_FLIGHT_TEST_LINKAGE "static" PARENT_SCOPE)
+endif()
+if (NOT ARROW_GRPC_USE_SHARED)
+  message(STATUS "Linking Arrow Flight tests statically due to static gRPC")
+  set(ARROW_FLIGHT_TEST_LINKAGE "static" PARENT_SCOPE)
+endif()
+
+
+if(ARROW_FLIGHT_TEST_LINKAGE STREQUAL "static")
   set(ARROW_FLIGHT_TEST_LINK_LIBS
       arrow_flight_static arrow_flight_testing_static ${ARROW_FLIGHT_STATIC_LINK_LIBS}
       ${ARROW_TEST_LINK_LIBS})
diff --git a/cpp/src/arrow/flight/flight_sql/CMakeLists.txt b/cpp/src/arrow/flight/flight_sql/CMakeLists.txt
index fd28559a9..f2bbb8f9b 100644
--- a/cpp/src/arrow/flight/flight_sql/CMakeLists.txt
+++ b/cpp/src/arrow/flight/flight_sql/CMakeLists.txt
@@ -36,7 +36,7 @@ set_source_files_properties(${FLIGHT_SQL_GENERATED_PROTO_FILES} PROPERTIES GENER
 
 add_custom_target(flight_sql_protobuf_gen ALL DEPENDS ${FLIGHT_SQL_GENERATED_PROTO_FILES})
 
-set(ARROW_FLIGHT_SQL_SRCS server.cc client.cc)
+set(ARROW_FLIGHT_SQL_SRCS server.cc client.cc "${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.cc")
 
 add_arrow_lib(arrow_flight_sql
               CMAKE_PACKAGE_NAME
@@ -58,7 +58,7 @@ add_arrow_lib(arrow_flight_sql
               STATIC_LINK_LIBS
               arrow_flight_static)
 
-if(ARROW_TEST_LINKAGE STREQUAL "static")
+if(ARROW_FLIGHT_TEST_LINKAGE STREQUAL "static")
   set(ARROW_FLIGHT_TEST_LINK_LIBS
       arrow_flight_static arrow_flight_testing_static ${ARROW_FLIGHT_STATIC_LINK_LIBS}
       ${ARROW_TEST_LINK_LIBS})
@@ -81,7 +81,6 @@ add_arrow_test(flight_sql_test
                SOURCES
                client_test.cc
                server_test.cc
-               "${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.cc"
                STATIC_LINK_LIBS
                ${ARROW_FLIGHT_SQL_TEST_LINK_LIBS}
                LABELS
@@ -97,8 +96,7 @@ if(ARROW_BUILD_TESTS OR ARROW_BUILD_EXAMPLES)
                  example/sqlite_statement.cc
                  example/sqlite_statement_batch_reader.cc
                  example/sqlite_server.cc
-                 example/sqlite_tables_schema_batch_reader.cc
-                 "${CMAKE_CURRENT_BINARY_DIR}/FlightSql.pb.cc")
+                 example/sqlite_tables_schema_batch_reader.cc)
   target_link_libraries(flight_sql_test_server
                         PRIVATE ${ARROW_FLIGHT_SQL_TEST_LINK_LIBS} ${GFLAGS_LIBRARIES}
                                 ${SQLite3_LIBRARIES})
```

Reply:
@lidavidm , this worked for me!